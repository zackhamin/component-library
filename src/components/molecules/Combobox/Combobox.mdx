# Combobox

This component is a combination of a text input and a dropdown, allowing users to filter a list of options to match a specific query.

It is particularly useful in cases where there are way too many option that would make a select difficult to use or for when there are more complex requirements that a select wouldn't satisfy.

The Combobox component is built on top of React Aria's implementation to ensure that all accessibility and edge cases are handled for us. You should not need to import anything directly from React Aria or associated libraries when using the Combobox component. All required components should be importable from our component library directly.

Please refer to the documention provided by React Aria for additional examples and functionality provided.

# How to use

## Non-Async

The Combobox can be used in the following way when dealing with non-async data.

```
    import { ComboBox, Item } from 'cc-component-library';

    <ComboBox label='Custom label'>
        <Item key="red panda">Red Panda</Item>
        <Item key="cat">Cat</Item>
        <Item key="dog">Dog</Item>
        <Item key="aardvark">Aardvark</Item>
        <Item key="kangaroo">Kangaroo</Item>
        <Item key="snake">Snake</Item>
	</ComboBox>
```

## Async

You can use the combobox with async data using the useAsyncList hook

```
import { ComboBox, Item, useAsyncList } from 'cc-component-library';

const StarWars = () => {
	const results = useAsyncList<{ name: string }>({
		async load({ signal, filterText }) {
			const res = await fetch(
				`https://swapi.dev/api/people/?search=${filterText}`,
				{ signal }
			);

			const json = await res.json();

			return {
				items: json.results,
				cursor: json.next,
			};
		},
	});

	return (
		<>
			<ComboBox
				label="Star Wars Character Search"
				items={results.items}
				inputValue={results.filterText}
				onInputChange={results.setFilterText}
				loadingState={results.loadingState}
				onLoadMore={results.loadMore}
			>
				{(item) => <Item key={item.name}>{item.name}</Item>}
			</ComboBox>
		</>
	);
};
```
