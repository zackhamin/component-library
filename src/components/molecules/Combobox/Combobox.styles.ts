import { colors } from 'Styles';
import styled from 'styled-components';

export const InputGroup = styled.div`
	display: inline-flex;
	width: 100%;
	overflow: hidden;
	border-radius: 4px;
	position: relative;

	&:focus-within {
		box-shadow: 0 0 0 3px ${colors.lime};
	}
`;

export const StyledInput = styled.input`
	width: 100%;
	appearance: none;
	padding: 6px 8px;
	font-size: 1.6rem;
	border: 1px solid ${colors.darkGrey};
	border-right: none;
	border-radius: 4px 0 0 4px;
	height: 4.8rem;

	&:focus {
		outline: none;
	}
`;

export const StyledComboButton = styled.button`
	appearance: none;
	border: none;
	background: ${colors.emerald};
	display: flex;
	justify-content: center;
	align-items: center;
	color: ${colors.white};
	width: 4.8rem;
`;

export const Wrapper = styled.div`
	position: relative;
`;

export const SpinnerWrapper = styled.span`
	position: absolute;
	right: 6rem;
	top: 50%;
	transform: translate(-50%, -50%);

	& > div {
		margin: 0;
	}
`;
