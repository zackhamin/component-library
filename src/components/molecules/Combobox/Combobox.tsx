import { Chevron } from 'Atoms/Icons/Chevron';
import type { ComboBoxProps } from '@react-types/combobox';
import { FormLabel } from 'Atoms/FormLabel';
import { ListBox } from './Listbox';
import { Loading } from '../../atoms/Loading';
import { Popover } from './Popover';
import React from 'react';
import {
	InputGroup,
	SpinnerWrapper,
	StyledComboButton,
	StyledInput,
	Wrapper,
} from './Combobox.styles';
import { Item, useAsyncList, useComboBoxState } from 'react-stately';
import { useButton, useComboBox, useFilter } from 'react-aria';

type LoadingState =
	| 'loading'
	| 'sorting'
	| 'loadingMore'
	| 'error'
	| 'idle'
	| 'filtering';

interface ComboBoxWithAsyncProps<T> extends ComboBoxProps<T> {
	loadingState?: LoadingState;
	onLoadMore?: () => void;
}

function ComboBox<T extends object>(props: ComboBoxWithAsyncProps<T>) {
	// Setup filter function and state.
	const { contains } = useFilter({ sensitivity: 'base' });
	const state = useComboBoxState({ ...props, defaultFilter: contains });

	// Setup refs and get props for child elements.
	const buttonRef = React.useRef(null);
	const inputRef = React.useRef(null);
	const listBoxRef = React.useRef(null);
	const popoverRef = React.useRef(null);

	const {
		buttonProps: triggerProps,
		inputProps,
		listBoxProps,
		labelProps,
	} = useComboBox(
		{
			...props,
			inputRef,
			buttonRef,
			listBoxRef,
			popoverRef,
		},
		state
	);

	const { buttonProps } = useButton(triggerProps, buttonRef);

	return (
		<Wrapper>
			<FormLabel {...labelProps}>{props.label}</FormLabel>
			<InputGroup>
				<StyledInput {...inputProps} ref={inputRef} />
				<StyledComboButton {...buttonProps} ref={buttonRef}>
					<Chevron aria-hidden="true" />
				</StyledComboButton>
				<SpinnerWrapper>
					{props.loadingState === 'loading' ||
					props.loadingState === 'filtering' ? (
						<Loading size="2.2rem" />
					) : null}
				</SpinnerWrapper>
			</InputGroup>
			{state.isOpen && (
				<Popover
					popoverRef={popoverRef}
					isOpen={state.isOpen}
					onClose={state.close}
				>
					<ListBox {...listBoxProps} listBoxRef={listBoxRef} state={state} />
				</Popover>
			)}
		</Wrapper>
	);
}

export { ComboBox, Item, useAsyncList };
