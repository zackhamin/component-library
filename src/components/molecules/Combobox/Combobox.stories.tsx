import { ComboBox, Item, useAsyncList } from './Combobox';
import { Meta, Story } from '@storybook/react';
import docs from './Combobox.mdx';

export default {
	title: 'molecules/Combobox',
	component: ComboBox,
	parameters: {
		docs: {
			page: docs,
		},
	},
} as Meta;

const Template: Story = (args) => {
	return (
		<>
			<ComboBox {...args}>
				<Item key="red panda">Red Panda</Item>
				<Item key="cat">Cat</Item>
				<Item key="dog">Dog</Item>
				<Item key="aardvark">Aardvark</Item>
				<Item key="kangaroo">Kangaroo</Item>
				<Item key="snake">Snake</Item>
			</ComboBox>
		</>
	);
};

export const Default = Template.bind({});
Default.args = {
	label: 'Custom label',
};

export const WithDefaultValue = Template.bind({});
WithDefaultValue.args = {
	label: 'Custom label',
	defaultInputValue: 'kangaroo',
};

export const DisabledItems = Template.bind({});
DisabledItems.args = {
	disabledKeys: ['kangaroo', 'cat'],
};

const AsyncTemplate: Story = () => {
	const results = useAsyncList<{ name: string }>({
		async load({ signal, filterText }) {
			const res = await fetch(
				`https://swapi.dev/api/people/?search=${filterText}`,
				{ signal }
			);

			const json = await res.json();

			return {
				items: json.results,
				cursor: json.next,
			};
		},
	});

	return (
		<>
			<ComboBox
				label="Star Wars Character Search"
				items={results.items}
				inputValue={results.filterText}
				onInputChange={results.setFilterText}
				loadingState={results.loadingState}
				onLoadMore={results.loadMore}
			>
				{(item) => <Item key={item.name}>{item.name}</Item>}
			</ComboBox>
		</>
	);
};

export const AsyncCombobox = AsyncTemplate.bind({});
