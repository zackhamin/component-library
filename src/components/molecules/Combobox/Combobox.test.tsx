import { OverlayContainer } from '@react-aria/overlays';
import userEvent from '@testing-library/user-event';
import { ComboBox, Item } from './Combobox';
import { fireEvent, render, screen } from '@testing-library/react';

const renderCombobox = () => {
	return render(
		<OverlayContainer>
			<ComboBox label="Test Heading">
				<Item key="red panda">Red Panda</Item>
				<Item key="cat">Cat</Item>
				<Item key="dog">Dog</Item>
				<Item key="aardvark">Aardvark</Item>
				<Item key="snake">Snake</Item>
			</ComboBox>
		</OverlayContainer>
	);
};

describe('Combobox', () => {
	it(' Should render correctly', () => {
		const { container } = render(
			<OverlayContainer>
				<ComboBox label="Test Heading">
					<Item key="red panda">Red Panda</Item>
					<Item key="cat">Cat</Item>
					<Item key="dog">Dog</Item>
					<Item key="aardvark">Aardvark</Item>
					<Item key="kangaroo">Kangaroo</Item>
					<Item key="snake">Snake</Item>
				</ComboBox>
			</OverlayContainer>
		);
		expect(container).toMatchSnapshot();
	});

	it('should display the correct number of options', () => {
		renderCombobox();

		const button = screen.getByRole('button');
		fireEvent.click(button);

		expect(screen.getAllByRole('option').length).toBe(5);
	});
});

describe('Listbox should show on click', () => {
	it(' Should render a listbox when button clicked', () => {
		renderCombobox();

		const button = screen.getByRole('button');

		fireEvent.click(button);

		expect(screen.getByRole('listbox')).toBeInTheDocument();
	});
});

describe('User Input', () => {
	it('should be possible for a user to select an option', async () => {
		renderCombobox();

		const button = screen.getByRole('button');
		const input = screen.getByRole('combobox');

		userEvent.click(button);
		userEvent.click(screen.getByText('Cat'));

		expect(input).toHaveValue('Cat');
	});

	it('should show listbox with option when user types in input', () => {
		renderCombobox();

		const input = screen.getByRole('combobox');

		userEvent.type(input, 'Ca');

		expect(screen.getByRole('listbox')).toBeInTheDocument();
		expect(screen.getByRole('option')).toHaveTextContent('Cat');
	});

	it('should clear input when invalid input added and enter pressed', () => {
		renderCombobox();

		const input = screen.getByRole('combobox');

		userEvent.type(input, 'X');
		userEvent.type(input, '{enter}');

		expect(input).toHaveValue('');
	});

	it('should clear input when invalid input added and click outside', () => {
		renderCombobox();

		const input = screen.getByRole('combobox');

		userEvent.type(input, 'X');
		userEvent.click(document.body);

		expect(input).toHaveValue('');
	});
});
