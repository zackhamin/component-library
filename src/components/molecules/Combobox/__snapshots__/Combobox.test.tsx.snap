// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox  Should render correctly 1`] = `<div />`;
